src/aws-cpp-sdk-core/include/smithy/identity/signer/built-in/SigV4aSigner.h

```diff
@@ -0,0 +1,202 @@
+
+            if(!success)
+            {
+                AWS_LOGSTREAM_ERROR(v4AsymmetricLogTag, "Failed to get Auth configureation");
+
+                return SigningError(Aws::Client::CoreErrors::MEMORY_ALLOCATION, "", "Failed to get Auth configureation", false);
+            }
+
+            std::shared_ptr<Aws::Crt::Http::HttpRequest> crtHttpRequest = request.ToCrtHttpRequest();
+
+            auto sigv4HttpRequestSigner = Aws::MakeShared<Aws::Crt::Auth::Sigv4HttpRequestSigner>(v4AsymmetricLogTag);
+            
+            sigv4HttpRequestSigner->SignRequest(crtHttpRequest, awsSigningConfig,
+                [&request, &success, this](const std::shared_ptr<Aws::Crt::Http::HttpRequest>& signedCrtHttpRequest, int errorCode) {
+                    success = (errorCode == AWS_ERROR_SUCCESS);
+                    if (success)
+                    {
+                        if (m_signatureType == Aws::Crt::Auth::SignatureType::HttpRequestViaHeaders)
+                        {
+                            for (size_t i = 0; i < signedCrtHttpRequest->GetHeaderCount(); i++)
+                            {
+                                Aws::Crt::Optional<Aws::Crt::Http::HttpHeader> httpHeader = signedCrtHttpRequest->GetHeader(i);
+                                request.SetHeaderValue(Aws::String(reinterpret_cast<const char*>(httpHeader->name.ptr), httpHeader->name.len),
+                                    Aws::String(reinterpret_cast<const char*>(httpHeader->value.ptr), httpHeader->value.len));
+                            }
+                        }
+                        else if (m_signatureType == Aws::Crt::Auth::SignatureType::HttpRequestViaQueryParams)
+                        {
+                            Aws::Http::URI newPath(reinterpret_cast<const char*>(signedCrtHttpRequest->GetPath()->ptr));
+                            request.GetUri().SetQueryString(newPath.GetQueryString());
+                        }
+                        else
+                        {
+                            AWS_LOGSTREAM_ERROR(v4AsymmetricLogTag, "No action to take when signature type is neither \"HttpRequestViaHeaders\" nor \"HttpRequestViaQueryParams\"");
+                            success = false;
+                        }
+                    }
+                    else
+                    {
+                        AWS_LOGSTREAM_ERROR(v4AsymmetricLogTag, "Encountered internal error during signing process with AWS signature version 4 (Asymmetric):" << aws_error_str(errorCode));
+                    }
+                }
+            );
+            
+            
+            return success? SigningFutureOutcome(std::move(httpRequest)) : SigningError(Aws::Client::CoreErrors::MEMORY_ALLOCATION, "", "Failed to sign the request with sigv4", false);
+        }
+
+
+
+        virtual ~AwsSigV4aSigner() {};
+    protected:
+
+        bool createAwsSigningConfig(
```

Comment:
why are we returning a bool, seems like we should returning a `Outcome<Aws::Crt::Auth::AwsSigningConfig, SomeError>`. if this was the way it was before might as well clean it up now.

Reply:
This is an internal helper, the output is the reference parameter. The real outcome is one step above

Reply:
i know the outcome is in a reference parameter, im saying instead of doing that we should create it here and return it via RVO/NVRO in a outcome